### PR TITLE
Standardized Contributions

### PR DESCRIPTION
**Pull request by wSedlacek**
fixes: #13 

This PR automates much of the contribution process

- [x] CI/CD for Testing and Linting
- [x] Commit Linting
- [x] Commitizen
- [x] Semantic Releases
- [x] Contribution guide
- [x] @kleydon Tag current master at `1.0.0` (to reset the versioning for semantic release)
- [x] @kleydon Setup GIT_TOKEN and NPM_TOKEN in repo secrets
- [x] @kleydon Protect Master branch for:
  - Require status checks to pass before merging
  - Require pull request reviews before merging
  - Include administrators

Note: Semantic release wants to use the previous tagged version then bump from there, to ensure that it uses `1.0.0` to start then the most recent tag needs to be `1.0.0`

Example of Release
https://github.com/wSedlacek/prisma-session-store/runs/1103231448?check_suite_focus=true

Example of Test/Lint
https://github.com/wSedlacek/prisma-session-store/runs/1103231421?check_suite_focus=true
https://github.com/wSedlacek/prisma-session-store/runs/1103231454?check_suite_focus=true

Example of PR checks
Screen Shot 2020-09-12 at 11 47 46

### CODE REVIEW

**Comment by wSedlacek** [comment]
Ready for review

**Comment by kleydon** [comment]
Hi @wSedlacek,

I've bumped the most recent tag for master to `v1.0.0`. (Should the conventional "v" be there - or is that problematic?)

I've also set added an `NPM_TOKEN` secret, but am a bit puzzled about what is needed re: the `GIT_TOKEN`.

According to [GitHubs docs](https://docs.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#permissions-for-the-github_token), there is a GIT**HUB**_TOKEN, which is _automatically_ created:

```
GitHub automatically creates a GITHUB_TOKEN secret to use in your workflow. 
You can use the GITHUB_TOKEN to authenticate in a workflow run.

When you enable GitHub Actions, GitHub installs a GitHub App on your repository. 
The GITHUB_TOKEN secret is a GitHub App installation access token. 
You can use the installation access token to authenticate on behalf of
the GitHub App installed on your repository.
```

Local and 3rd-party actions are already enabled for this repo, so I suspect a GITHUB_TOKEN is _already_ created (though it doesn't appear in the repo's "secrets" settings... Not sure why).

Is this `GITHUB_TOKEN` what is necessary? Or is a separate **GIT**_TOKEN required (as you wrote)?

If a separate `GIT_TOKEN` is required: What permissions are absolutely necessary to include? I'm guessing they are from [this list](https://docs.github.com/en/enterprise/2.15/user/articles/creating-a-personal-access-token-for-the-command-line)?

Let me know,

-K

**Comment by kleydon** [comment]
Ah - I see in the code it is GIT_TOKEN (as you specified). Can you let me know what permissions are necessary for this to work?

File: README.md
Comment (kleydon) [inline README.md:187]: Consider "...or wherever you would like, even..."

File: src/lib/utils/create-expiration.ts
Comment (kleydon) [inline src/lib/utils/create-expiration.ts:6]: This is nitpicky, but does it make sense for rounding to be _any_ number, or a _limited set of numbers_ (e.g. 10, 100, 1000 ms)? Can typescript handle this, via 10 | 100 | 1000, etc? Alternatively: Might it make sense for the parameter to be a number corresponding to a place-value, or a precision, rather than a number of milliseconds? (Just offering food for thought; trusting you'll pick the approach that makes the most sense...)

**Comment by kleydon** [comment]
Ok, I've added a `GIT_TOKEN`; if it turns out not to be permissive enough for the CI/CD tasks, let me know.

**Comment by wSedlacek** [comment]
> Is this `GITHUB_TOKEN` what is necessary? Or is a separate **GIT**_TOKEN required (as you wrote)?
> 
> If a separate `GIT_TOKEN` is required: What permissions are absolutely necessary to include? I'm guessing they are from [this list](https://docs.github.com/en/enterprise/2.15/user/articles/creating-a-personal-access-token-for-the-command-line)?

Yes it is a Personal Access token to GitHub with public_repo access.
This is what allows the semantic release to push to GitHub (the GITHUB_TOKEN provided by the workflow does not have push access)

Screen Shot 2020-09-12 at 18 38 14

I won't know for sure if it is correct until this is merged and able to run.

**Comment by wSedlacek** [comment]
> I've bumped the most recent tag for master to `v1.0.0`. (Should the conventional "v" be there - or is that problematic?)

That should work.

**Review by kleydon — APPROVED** [review]

**Comment by kleydon** [comment]
>Yes it is a Personal Access token to GitHub with public_repo access.
>This is what allows the semantic release to push to GitHub (the GITHUB_TOKEN provided by the workflow does not have push access)
>Update: Here is the source for further information.

Thanks.

Ok; I've tagged to v1.0.0, set up the tokens, and protected the master branch. Going to merge, and we'll see what happens!

If there's anything that needs to be fiddled with on my end - let me know; curious to see this new level of sophistication/automation in action...

**Comment by kleydon** [comment]
Whoa - its alive! (Or at least it is doing things automagically...)

**Comment by wSedlacek** [comment]
> Whoa - its alive! (Or at least it is doing things automagically...)

Looks good! Should be ready to merge and we will see if it releases.

**Comment by kleydon** [comment]
Post merging, how does one monitor release progress?

**Comment by wSedlacek** [comment]
> Post merging, how does one monitor release progress?

In the Actions section.
https://github.com/kleydon/prisma-session-store/runs/1107279105

Looks like it didn't like the protected rules. Let's remove `Include administrators` then rerun the job.
Screen Shot 2020-09-12 at 19 00 32

**Comment by kleydon** [comment]
Removed `Include administrators` and re-running...

**Comment by wSedlacek** [comment]
> Removed `Include administrators` and re-running...

Looks like we are getting closer.
Seems like our NPM Token needs to be updated now.

**Comment by kleydon** [comment]
>>    'npm ERR! 403 403 Forbidden - PUT https://registry.npmjs.org/@quixo3%2fprisma-session-store - You may not perform that action with these credentials.\n' +
591

**Comment by wSedlacek** [comment]
> > > 'npm ERR! 403 403 Forbidden - PUT https://registry.npmjs.org/@quixo3%2fprisma-session-store - You may not perform that action with these credentials.\n' +
> > > 591

Screen Shot 2020-09-12 at 19 12 21
The token needs read and publish writes.

**Comment by kleydon** [comment]
>>Before you rerun remove the v1.1.0 tag that it published to avoid conflicts.

Done.

Tending to the NPM token now...

**Comment by wSedlacek** [comment]
> > > Before you rerun remove the v1.1.0 tag that it published to avoid conflicts.
> 
> Done.
> 
> Tending to the NPM token now...

Once we get this fixed up we will probably have duplicates in our CHANGELOG.md, I will make a quick PR to remove those duplicates.

**Comment by kleydon** [comment]
Ok; NPM token fixed.
Does a new version number need to be set to anything in particular?

**Comment by wSedlacek** [comment]
It should be fine to rerun as long as the tag was removed. I am slightly conserved that when we rerun we will be one commit behind but I think it will just stop the workflow if anything.

I would retry real quick and if it fails we can make a revert of the release commit.

**Comment by wSedlacek** [comment]
Yep, looks like a Revert Commit is needed.
https://github.com/kleydon/prisma-session-store/commit/239b192378944f37a4d082adb88e9fcffdb8ce04

Think you can take care of that?

**Comment by kleydon** [comment]
Will try here...

**Comment by wSedlacek** [comment]
If you run into any trouble I made a PR for you
https://github.com/kleydon/prisma-session-store/pull/15

**Comment by wSedlacek** [comment]
@kleydon Sorry about all the confusion.
I know CI/CD can be pretty finicky when you are first getting it setup, once it is setup though it should be pretty solid.

**Comment by kleydon** [comment]
Looks like something went right? ^

**Comment by wSedlacek** [comment]
🎉  We did it!

**Comment by wSedlacek** [comment]
Anyways, I think this library is pretty solid now.
Gonna shift back to working on some of my other projects but feel free to @me if you need my help with anything.

**Comment by kleydon** [comment]
It still LOOKED like the v1.0.0 tag didn't get removed (but perhaps I was looking at the recreated version...)
At any rate - progress.
Thanks for your patience (and great job on setting up the "machinery" itself...)
Ok - enjoy!

**Comment by wSedlacek** [comment]
> It still LOOKED like the v1.0.0 tag didn't get removed (but perhaps I was looking at the recreated version...)
> At any rate - progress.
> Thanks for your patience (and great job on setting up the "machinery" itself...)
> Ok - enjoy!

Well the base tag before the release was `v1.0.0` so the release was `v1.1.0` which as far as I can tell went perfect.
Note: If we started at `v0.9.0` we would be at `v0.10.0` now. Didn't really know of a good way to get us to `1.0.0` so I figured `v1.1.0` was the next best thing.

**Comment by kleydon** [comment]
Hey @wSedlacek,

Just sent a contributor invite. 

Also, wanted to mention that if you ever want to swap roles, such that you take ownership of the repo and I'm a contributor - just let me know. You've done all this great work, and I don't want you to feel like you aren't getting proper credit for it.

**Comment by wSedlacek** [comment]
Thank you for the invite.

It's all good. In all honestly I prefer not to be the owner as I have a lot of projects I am responsible for already. My work on this project benefited me a lot as I am using it in a few other projects so I am just happy I was able to get this working the way I like and share my work with others so they can benefit from it as well. 

Again thank you for the invite very happy to accept and help if any other issues need to be taken care of.

**Comment by kleydon** [comment]
Ok! Just figured I'd check. -K

**Pull request merged**
